Change the type of `NodeID` to `ulong`

linked0 opened the pull request:
This contains the code changing the type of `NodeID` to `ulong` and removing quorum hashes from SCP protocol messages.

Fixes #1191, #982

linked0 committed:
Add an overload to vector.push_back that accepts scalars

This allow us to push a 'const' scalar back,
as it can be copied to a non-const instance.

linked0 commented:
This PR has some link errors that I need to deal with and I'd like to have a white-boarding about if necessary.

```
Linking...
Undefined symbols for architecture x86_64:
  "bool cpp_set_empty<unsigned long>(void const*)", referenced from:
      __D4scpd3Cpp__T3setTmTSQuQr__T4lessTmZQiTSQBnQBl__T9allocatorTmZQnZQCa5emptyMxFNaNbNiZb in agora.o
  "void dtorCPPObject<std::__1::unordered_map<unsigned long, stellar::QuorumTracker::NodeInfo, stellar::RandHasher<unsigned long, std::__1::hash<unsigned long> >, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, stellar::QuorumTracker::NodeInfo> > > >(std::__1::unordered_map<unsigned long, stellar::QuorumTracker::NodeInfo, stellar::RandHasher<unsigned long, std::__1::hash<unsigned long> >, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, stellar::QuorumTracker::NodeInfo> > >*)", referenced from:
      __D4scpd3Cpp__T13unordered_mapTmTSQBf6quorum13QuorumTrackerQp8NodeInfoTSQCrQCp__T10RandHasherTmTSQDqQDo__T4hashTmZQiZQBkTSQEpQEn__T8equal_toTmZQmTSQFoQFm__T9allocatorTSQGjQGh__T4pairTxmTQFxZQnZQBlZQHa8__mixin46__dtorMFNbNiNeZv in agora.o
  "void dtorCPPObject<std::__1::map<unsigned long, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper>, std::__1::less<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper> > > > >(std::__1::map<unsigned long, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper>, std::__1::less<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper> > > >*)", referenced from:
      __D4scpd3Cpp__T3mapTmTSQuQr__T10shared_ptrTSQBp3scp9SCPDriver18SCPEnvelopeWrapperZQCaTSQDgQDe__T4lessTmZQiTSQEbQDz__T9allocatorTSQEwQEu__T4pairTxmTQEvZQnZQBlZQFn8__mixin26__dtorMFNbNiNeZv in agora.o
  "void dtorCPPObject<std::__1::set<unsigned long, std::__1::less<unsigned long>, std::__1::allocator<unsigned long> > >(std::__1::set<unsigned long, std::__1::less<unsigned long>, std::__1::allocator<unsigned long> >*)", referenced from:
      __D4scpd3Cpp__T3setTmTSQuQr__T4lessTmZQiTSQBnQBl__T9allocatorTmZQnZQCa8__mixin46__dtorMFNbNiNeZv in agora.o
  "int cpp_set_foreach<unsigned long>(void*, void*, void*)", referenced from:
      __D4scpd3Cpp__T3setTmTSQuQr__T4lessTmZQiTSQBnQBl__T9allocatorTmZQnZQCa7opApplyMxFMDFKxmZiZi in agora.o
  "void copyCtorCPPObject<std::__1::unordered_map<unsigned long, stellar::QuorumTracker::NodeInfo, stellar::RandHasher<unsigned long, std::__1::hash<unsigned long> >, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, stellar::QuorumTracker::NodeInfo> > > >(std::__1::unordered_map<unsigned long, stellar::QuorumTracker::NodeInfo, stellar::RandHasher<unsigned long, std::__1::hash<unsigned long> >, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, stellar::QuorumTracker::NodeInfo> > >*, std::__1::unordered_map<unsigned long, stellar::QuorumTracker::NodeInfo, stellar::RandHasher<unsigned long, std::__1::hash<unsigned long> >, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, stellar::QuorumTracker::NodeInfo> > >*)", referenced from:
      __D4scpd3Cpp__T13unordered_mapTmTSQBf6quorum13QuorumTrackerQp8NodeInfoTSQCrQCp__T10RandHasherTmTSQDqQDo__T4hashTmZQiZQBkTSQEpQEn__T8equal_toTmZQmTSQFoQFm__T9allocatorTSQGjQGh__T4pairTxmTQFxZQnZQBlZQHa8__mixin46__ctorMFNbNcNiNeKNgSQItQIr__TQIqTmTQIeTQGwTQFcTQEhZQJmZSQKdQKb__TQKaTmTQJoTQIgTQGmTQFrZQKw in agora.o
      __D4scpd3Cpp__T13unordered_mapTmTSQBf6quorum13QuorumTrackerQp8NodeInfoTSQCrQCp__T10RandHasherTmTSQDqQDo__T4hashTmZQiZQBkTSQEpQEn__T8equal_toTmZQmTSQFoQFm__T9allocatorTSQGjQGh__T4pairTxmTQFxZQnZQBlZQHa8__mixin46__ctorMyFNbNcNiNeKNgSQIuQIs__TQIrTmTQIfTQGxTQFdTQEiZQJnZySQKfQKd__TQKcTmTQJqTQIiTQGoTQFtZQKy in agora.o
  "void copyCtorCPPObject<std::__1::map<unsigned long, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper>, std::__1::less<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper> > > > >(std::__1::map<unsigned long, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper>, std::__1::less<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper> > > >*, std::__1::map<unsigned long, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper>, std::__1::less<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper> > > >*)", referenced from:
      __D4scpd3Cpp__T3mapTmTSQuQr__T10shared_ptrTSQBp3scp9SCPDriver18SCPEnvelopeWrapperZQCaTSQDgQDe__T4lessTmZQiTSQEbQDz__T9allocatorTSQEwQEu__T4pairTxmTQEvZQnZQBlZQFn8__mixin26__ctorMFNbNcNiNeKNgSQHgQHe__TQHdTmTQHcTQEuTQEdZQHvZSQImQIk__TQIjTmTQIiTQGaTQFjZQJb in agora.o
      __D4scpd3Cpp__T3mapTmTSQuQr__T10shared_ptrTSQBp3scp9SCPDriver18SCPEnvelopeWrapperZQCaTSQDgQDe__T4lessTmZQiTSQEbQDz__T9allocatorTSQEwQEu__T4pairTxmTQEvZQnZQBlZQFn8__mixin26__ctorMyFNbNcNiNeKNgSQHhQHf__TQHeTmTQHdTQEvTQEeZQHwZySQIoQIm__TQIlTmTQIkTQGcTQFlZQJd in agora.o
  "void copyCtorCPPObject<std::__1::set<unsigned long, std::__1::less<unsigned long>, std::__1::allocator<unsigned long> > >(std::__1::set<unsigned long, std::__1::less<unsigned long>, std::__1::allocator<unsigned long> >*, std::__1::set<unsigned long, std::__1::less<unsigned long>, std::__1::allocator<unsigned long> >*)", referenced from:
      __D4scpd3Cpp__T3setTmTSQuQr__T4lessTmZQiTSQBnQBl__T9allocatorTmZQnZQCa8__mixin46__ctorMFNbNcNiNeKNgSQDtQDr__TQDqTmTQDpTQDaZQEeZSQEvQEt__TQEsTmTQErTQEcZQFg in agora.o
      __D4scpd3Cpp__T3setTmTSQuQr__T4lessTmZQiTSQBnQBl__T9allocatorTmZQnZQCa8__mixin46__ctorMyFNbNcNiNeKNgSQDuQDs__TQDrTmTQDqTQDbZQEfZySQExQEv__TQEuTmTQEtTQEeZQFi in agora.o
  "void defaultCtorCPPObject<std::__1::unordered_map<unsigned long, stellar::QuorumTracker::NodeInfo, stellar::RandHasher<unsigned long, std::__1::hash<unsigned long> >, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, stellar::QuorumTracker::NodeInfo> > > >(std::__1::unordered_map<unsigned long, stellar::QuorumTracker::NodeInfo, stellar::RandHasher<unsigned long, std::__1::hash<unsigned long> >, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, stellar::QuorumTracker::NodeInfo> > >*)", referenced from:
      __D4scpd3Cpp__T13unordered_mapTmTSQBf6quorum13QuorumTrackerQp8NodeInfoTSQCrQCp__T10RandHasherTmTSQDqQDo__T4hashTmZQiZQBkTSQEpQEn__T8equal_toTmZQmTSQFoQFm__T9allocatorTSQGjQGh__T4pairTxmTQFxZQnZQBlZQHa8__mixin46__ctorMFNbNcNiNeEQIqQIo7CppCtorZSQJgQJe__TQJdTmTQIrTQHjTQFpTQEuZQJz in agora.o
  "void defaultCtorCPPObject<std::__1::map<unsigned long, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper>, std::__1::less<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper> > > > >(std::__1::map<unsigned long, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper>, std::__1::less<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, std::__1::shared_ptr<stellar::SCPEnvelopeWrapper> > > >*)", referenced from:
      __D4scpd3Cpp__T3mapTmTSQuQr__T10shared_ptrTSQBp3scp9SCPDriver18SCPEnvelopeWrapperZQCaTSQDgQDe__T4lessTmZQiTSQEbQDz__T9allocatorTSQEwQEu__T4pairTxmTQEvZQnZQBlZQFn8__mixin26__ctorMFNbNcNiNeEQHdQHb7CppCtorZSQHtQHr__TQHqTmTQHpTQFhTQEqZQIi in agora.o
  "void defaultCtorCPPObject<std::__1::set<unsigned long, std::__1::less<unsigned long>, std::__1::allocator<unsigned long> > >(std::__1::set<unsigned long, std::__1::less<unsigned long>, std::__1::allocator<unsigned long> >*)", referenced from:
      __D4scpd3Cpp__T3setTmTSQuQr__T4lessTmZQiTSQBnQBl__T9allocatorTmZQnZQCa8__mixin46__ctorMFNbNcNiNeEQDqQDo7CppCtorZSQEgQEe__TQEdTmTQEcTQDnZQEr in agora.o
  "void cpp_unordered_map_assign<unsigned long, stellar::QuorumTracker::NodeInfo>(void*, unsigned long const&, stellar::QuorumTracker::NodeInfo const&)", referenced from:
      std::__1::unordered_map<unsigned long, stellar::QuorumTracker::NodeInfo, stellar::RandHasher<unsigned long, std::__1::hash<unsigned long> >, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, stellar::QuorumTracker::NodeInfo> > >::opIndexAssign(stellar::QuorumTracker::NodeInfo const&, unsigned long) in agora.o
  "unsigned long cpp_unordered_map_length<unsigned long, stellar::QuorumTracker::NodeInfo>(void const*)", referenced from:
      std::__1::unordered_map<unsigned long, stellar::QuorumTracker::NodeInfo, stellar::RandHasher<unsigned long, std::__1::hash<unsigned long> >, std::__1::equal_to<unsigned long>, std::__1::allocator<std::__1::pair<unsigned long const, stellar::QuorumTracker::NodeInfo> > >::length() in agora.o
  "void push_back<unsigned long const, xdr::xvector<unsigned long, 4294967292u> >(xdr::xvector<unsigned long, 4294967292u>&, unsigned long const&)", referenced from:
      __D4scpd5types5Utils3dupFKxSQzQw11Stellar_SCP12SCPQuorumSetZSQCgQCeQBjQz in agora.o
  "void push_back<unsigned long, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > >(std::__1::vector<unsigned long, std::__1::allocator<unsigned long> >&, unsigned long&)", referenced from:
      __D4scpd3Cpp__T6vectorTmTSQxQu__T9allocatorTmZQnZQBi9push_backMFNaNbNiNeKmZv in agora.o
      __D4scpd3Cpp__T6vectorTmTSQxQu__T9allocatorTmZQnZQBi9push_backMFNaNbNiNemZv in agora.o
  "stellar::normalizeQSet(stellar::SCPQuorumSet&, unsigned long const*)", referenced from:
      __D5agora9consensus8protocol9NominatorQk14buildSCPConfigFNbNfIKSQCj6common5Types12QuorumConfigZ9__lambda3MFNbNiNeZv in agora.o
  "stellar::SCPDriver::toShortString(xdr::opaque_array<64u> const&) const", referenced from:
      vtable for stellar::SCPDriver in SCPDriver.o
  "stellar::SCPDriver::toStrKey(xdr::opaque_array<64u> const&, bool) const", referenced from:
      vtable for stellar::SCPDriver in SCPDriver.o
ld: symbol(s) not found for architecture x86_64
```

Geod24 commented on source/scpp/src/scp/SCPDriver.h:
```diff
@@ -100,6 +100,14 @@ class SCPDriver
     // should be flooded to the network.
     virtual void emitEnvelope(SCPEnvelope const& envelope) = 0;
 
+    // `getNodeID` gets the UTXO hash for a node with the `node_index`
+    // at the height that is handed in as `slotIndex`.
+    virtual Hash getNodeID(uint64 slotIndex, uint64 nodeIndex) = 0;
+
+    // `getNodeIndex` gets the index of the UTXO hash for a node with
+    // the `node_id` at the height that is handed in as `slotIndex`.
+    virtual uint64 getNodeIndex(uint64 slotIndex, Hash nodeID) = 0;
+
```
I don't see this being used ?

linked0 replied:
I made it for the case where a `NominationProtocol` could refer to a `NodeID` or an index. But will remove if unnecessary.

Geod24 commented on source/scpp/src/scp/SCP.h:
```diff
@@ -25,7 +25,7 @@ class SCP
 
   public:
-    SCP(SCPDriver& driver, NodeID const& nodeID, bool isValidator,
+    SCP(SCPDriver& driver, Hash const& nodeID, bool isValidator,
```
That line shouldn't be modified.

linked0 replied:
Reverted.

Geod24 commented on source/scpp/src/scp/SCP.cpp:
```diff
@@ -18,7 +18,7 @@
 {
 
-SCP::SCP(SCPDriver& driver, NodeID const& nodeID, bool isValidator,
+SCP::SCP(SCPDriver& driver, Hash const& nodeID, bool isValidator,
```
Ditto, shouldn't be modified

linked0 replied:
Oh, I'll revert it. There might be some mistakes because I'm in the middle of modifying.

Geod24 commented on source/scpd/types/Stellar_types.d:
```diff
@@ -44,6 +44,6 @@ unittest
 
     static assert(agora.crypto.Hash.Hash.sizeof == Hash.sizeof);
-    static assert(agora.crypto.Hash.Hash.sizeof == NodeID.sizeof);
+    static assert(ulong.sizeof == NodeID.sizeof);
```
You can remove the line, the other ones are to ensure binary compatibility, but since we use `ulong` directly this is not necessary.

linked0 replied:
Removed

Geod24 commented on source/scpd/scp/SCPDriver.d:
```diff
@@ -140,8 +150,9 @@ nothrow:
         {
-            auto slice = Hash(pk).toString();
-            return sliceToStdString(slice.ptr, slice.length);
+            // auto slice = Hash(pk).toString();
+            // return sliceToStdString(slice.ptr, slice.length);
+            return std_string("NODEID");
```
For now you can get away with:
```
auto slice = pk.to!string;
return sliceToStdString(slice.ptr, slice.length);
```

linked0 replied:
I'm getting away with it now.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -232,14 +232,16 @@ extern(D):
             if (this.scp is null && self_enroll != Hash.init)
             {
-                auto node_id = NodeID(self_enroll[][0 .. NodeID.sizeof]);
+                auto node_id = StellarHash(self_enroll[][0 .. StellarHash.sizeof]);
```
That's not correct either

linked0 replied:
Updated and will modify it more.

linked0 commented:
Only one error remains which I will fix soon.
```
In file included from /Users/hyunjaelee/work/agora/source/scpp/src/quorum/QuorumIntersectionCheckerImpl.cpp:8:
/Users/hyunjaelee/work/agora/source/scpp/src/crypto/KeyUtils.h:25:32: error: reference to type 'const stellar::ByteSlice' could not bind to an lvalue of type 'const unsigned long long'
    return strKey::toStrKey(0, key).value;
                               ^~~
/Users/hyunjaelee/work/agora/source/scpp/src/crypto/KeyUtils.h:31:12: note: in instantiation of function template specialization 'stellar::KeyUtils::toStrKey<unsigned long long>' requested here
    return toStrKey(key).substr(0, 5);
           ^
/Users/hyunjaelee/work/agora/source/scpp/src/quorum/QuorumIntersectionCheckerImpl.cpp:697:31: note: in instantiation of function template specialization 'stellar::KeyUtils::toShortString<unsigned long long>' requested here
    return stellar::KeyUtils::toShortString(mBitNumPubKeys.at(node));
                              ^
/Users/hyunjaelee/work/agora/source/scpp/src/crypto/StrKey.h:27:52: note: passing argument to parameter 'bin' here
SecretValue toStrKey(uint8_t ver, ByteSlice const& bin);
                                                   ^
```

linked0 commented:
I'm sorry but is there any simple way to covert a `unsigned long` to a `ByteSlice`? The previous error occurs in the code of `QuorumIntersectionCheckerImpl.cpp:697` as follows.
```
return stellar::KeyUtils::toShortString(mBitNumPubKeys.at(node));
```

linked0 commented:
With the support of @Geod24 about the mismatch `NodeID` and `ByteSlice`(e6bb6ae7a719da9fdb43c91902b6aa794e34e54f), I can so much progress in this PR. But there's one more link error about the bridge functions defined in the `DUtils.cpp` file.
I tried to modify the file for the link error and fail to solve the error. Can't adding `CPPSIZEOFINST(uint512);` solve the error?
```
Undefined symbols for architecture x86_64:
  "cppSizeOf(xdr::opaque_array<64u>&)", referenced from:
      __D4scpd5tests8SizeTest17__unittest_L26_C1FZv in agora-unittests.o
ld: symbol(s) not found for architecture x86_64
```

omerfirmak commented:
> Can't adding CPPSIZEOFINST(uint512); solve the error?

AFAICS, it should

Geod24 commented on source/scpp/src/scp/LocalNode.cpp:
```diff
@@ -99,7 +99,7 @@ LocalNode::computeWeight(uint64 m, uint64 total, uint64 threshold)
 // weight of the first occurrence
 uint64
-LocalNode::getNodeWeight(NodeID const& nodeID, SCPQuorumSet const& qset)
+LocalNode::getNodeWeight(NodeID const& nodeID, Hash const& nodeHash, SCPQuorumSet const& qset)
```
Why ?

linked0 replied:
I wanted the algorithm calculating the weight to be as it was, which uses the hash of UTXO.

Geod24 replied:
No it does not

linked0 replied:
I'll check it again.

Geod24 replied:
The parameter you added, `nodeHash`, is not used anywhere, is it ?

linked0 replied:
Removed the unnecessary parameter.

Geod24 commented on source/scpp/src/scp/QuorumSetUtils.cpp:
```diff
@@ -134,14 +134,14 @@ namespace
 
 void
-normalizeQSetSimplify(SCPQuorumSet& qSet, NodeID const* idToRemove)
+normalizeQSetSimplify(SCPQuorumSet& qSet, bool remove, NodeID const idToRemove)
```
I don't see why those changes are necessary either ?

linked0 replied:
Previously, the default value of `idToRemove` was `nullptr`, so it can be used to determine whether to remove the node or not. But we can not call this function in the situation where there's no node to be removed because the type `NodeID` is changed into `ulong`.

Geod24 replied:
Why not ? It passes a point to a `NodeID`, I don't see what's different here ?

linked0 replied:
Right, we don't need the new parameter, I'll check it out again.

linked0 committed:
Add function to change `NodeID` of validator

We change the type of `NodeID` to ulong which will be an index of frozen UTXO at the
certain height. So, the `NodeID` will be changed if there is an update in the validator
set. The function, `changeNodeID` is for the situation.

linked0 committed:
Change type of NodeID to ulong

SCP statements previously contain a `node_id` which is 64-byte key.
However because we are using staking we had the opportunity to
optimize that and reduce the network load. So we replace `node_id`
with a 8-byte `ulong` index. This index could be used to look up
the frozen UTXO of a validator for a specified height.

linked0 commented:
The modification of `source/scpd/tests/GlueTypes.d ` should be restored. But without the modification, this PR has a link error. This could be solved in the middle of reviews, I think.
```
Undefined symbols for architecture x86_64:
  "cppSizeOf(xdr::opaque_array<64u>&)", referenced from:
      __D4scpd5tests8SizeTest17__unittest_L26_C1FZv in agora-unittests.o
ld: symbol(s) not found for architecture x86_64
```

linked0 commented:
I have completed applying the code that removes quorum hashes from SCP protocol messages from [the PR #1192](https://github.com/bosagora/agora/pull/1192), which solves issue #982. I'm working on refining all the codes in commits in this PR.

linked0 commented:
Ready for review! :)

Geod24 commented on source/agora/common/Config.d:
```diff
@@ -1007,7 +1006,7 @@ public QuorumConfig toQuorumConfig (const ref SCPQuorumSet scp_quorum)
     import scpd.types.Stellar_types : NodeID;
 
-    Hash[] nodes;
+    ulong[] nodes;
```
NodeID

linked0 replied:
I'd like to remain as it is because there's some problem when sorting the node in Quorum.d.
```
quorum.nodes.sort;
```

linked0 replied:
I tried it several ways. But there are a couple of things to handle after changing that. So I think it would be OK to remain as it is.
The problems:
1. Serializer/Deserializer should be changed.
2. New code is needed to sort the array of `NodeID`.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -228,24 +228,33 @@ extern(D):
     ***************************************************************************/
 
-    public void createSCPObject () nothrow @safe
+    public void createSCPObject (in Height height) nothrow @safe
```
Probably needs a different name, like `updateSCPObject` ?

linked0 replied:
Updated.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1081,6 +1096,123 @@ extern(D):
+            return uint64_t.max;
+        }
+
+        Block block;
+        try
+        {
+            block = this.ledger.getBlocksFrom(Height(0))[0];
+        }
+        catch (Exception exc)
+        {
+            log.fatal("Exception happened with calling `getBlocksFrom` on `getNodeIndex: {}", exc);
+            assert(0);
+        }
```
That will only work for initially enrolled validator, will it not ?

linked0 replied:
Removed `getSelfUTXOIndex`.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1081,6 +1096,123 @@ extern(D):
+            assert(0);
+        }
+
+        const node_hash = StellarHash(self_enroll[][0 .. StellarHash.sizeof]);
+        if (slot_index == 0)
+        {
+            foreach (idx, enroll; block.header.enrollments)
+            {
+                if (enroll.utxo_key == node_hash)
+                    return idx;
+            }
+            log.fatal("Could not find a UTXO for this node at Genesis");
+            assert(0);
+        }
```
The only call I can see is at L243 in validator, which means we will never call this with `slot_index == 0`.

linked0 replied:
This is also called at L160 in `Validator.d`.

linked0 replied:
Ditto

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -228,24 +228,33 @@ extern(D):
             // enrolled with another UTXO after its previous enrollment expired.
-            auto self_enroll = this.enroll_man.getEnrollmentKey();
-            if (this.scp is null && self_enroll != Hash.init)
+            const node_id = this.getSelfUTXOIndex(height + 1);
```
Looking at the call chain, I think the caller of `createSCPObject` already has all the infos needed.

linked0 replied:
I don't think we need `getSelfUTXOIndex` with your comment. I removed it.

Geod24 reviewed (changes requested):
Need to review the last commit, but first 3 LGTM, comments mostly on the 4th.
Last commit also includes an openssl downgrade which was committed accidentally I suppose.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -98,6 +98,7 @@ public extern (C++) class Nominator : SCPDriver
 
     /// The mapping of all known quorum sets
     private SCPQuorumSetPtr[Hash] known_quorums;
+    private SCPQuorumSetPtr[NodeID] all_quorums;
```
Why the new member ? Isn't it replacing `known_quorums` ?

linked0 replied:
Reverted to `known_quorums`.

linked0 committed:
Remove quorum hashes from SCP protocol messages

The quorum hashes can be implicitly derived from the `node_id` and
`slot_idx` fields which are part of every `SCPEnvelope` message.
This reduces the total network overhead.

Geod24 commented on source/scpp/extra/DSCPUtils.cpp:
```diff
@@ -9,7 +9,7 @@
 using namespace stellar;
 
-SCP* createSCP(SCPDriver* driver, NodeID const& nodeID, bool isValidator, SCPQuorumSet const& qSetLocal)
+SCP* createSCP(SCPDriver* driver, NodeID nodeID, bool isValidator, SCPQuorumSet const& qSetLocal)
```
Why the change ?

linked0 replied:
It generated an error like this.
```
Linking...
Undefined symbols for architecture x86_64:
  "createSCP(stellar::SCPDriver*, unsigned long long, bool, stellar::SCPQuorumSet const&)", referenced from:
      __D5agora9consensus8protocol9NominatorQk15updateSCPObjectMFNbNfISQCk6common5Types6HeightIE4core4stdc6config13__c_ulonglongZ9__lambda5MFNeZv in agora-unittests.o
ld: symbol(s) not found for architecture x86_64
```

Geod24 commented on source/scpd/quorum/QuorumIntersectionChecker.d:
```diff
@@ -19,6 +19,17 @@ import scpd.Cpp;
 import scpd.quorum.QuorumTracker;
 import scpd.types.Stellar_types;
 
+extern (C++, "_GLOBAL__N_1") {
```
DDOC

linked0 replied:
I'm sorry but I have no idea how I DDOC this. Could you let me know the right comment for that?

Geod24 replied:
```
// Quorum intersection checker implementation
// This method was originally implemented in C++ but since the changes to `NodeID` had to be moved to D.
```

Something like this.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -1081,6 +1103,61 @@ extern(D):
+        @safe nothrow
+    {
+        Hash utxo;
+        if (slot_index == 1)
+        {
+            Block b0;
+            try
+            {
+                b0 = this.ledger.getBlocksFrom(Height(0))[0];
+            }
+            catch (Exception exc)
+            {
+                log.error("Exception happened with getting block at requested height: {}", exc);
+            }
+            assert(b0.header.enrollments.length > index);
+            utxo = b0.header.enrollments[index].utxo_key;
+        }
+        else
```
Why is `slot_index == 1` treated specially ?

linked0 replied:
Removed it.

Geod24 reviewed (commented):
Nice. I'll give it a shot soon, but besides the one comment about the branch, LGTM.

Geod24 commented on source/scpp/src/scp/BallotProtocol.cpp:
```diff
@@ -1564,27 +1561,6 @@ BallotProtocol::hasPreparedBallot(SCPBallot const& ballot,
-    case SCP_ST_CONFIRM:
-        h = st.pledges.confirm().quorumSetHash;
-        break;
-    case SCP_ST_EXTERNALIZE:
-        h = st.pledges.externalize().commitQuorumSetHash;
-        break;
```
Do you know why Stellar had this difference ?

linked0 replied:
For the line, I don't know the reason which I should analyze more about.

linked0 replied:
The `commitQuorumSetHash` is the same value as `quorumSetHash` of other types of SCPStatement. It's never changed as long as the quorum set remains the same. So it should be removed.

https://github.com/bosagora/agora/blob/bfe95e853eef28f8487d9a2fc8897b86d2852554/source/scpp/src/scp/BallotProtocol.cpp#L560-L576

Geod24 reviewed (approved):
LGTM, just one question.
